primitives/runtime/src/traits.rs

```diff
@@ -1349,6 +1349,47 @@ pub trait BenchmarkingSetup<T, Call, RawOrigin> {
 	fn instance(&self, components: &[(BenchmarkParameter, u32)]) -> Result<(Call, RawOrigin), &'static str>;
 }
 
+/// Implement `BenchmarkingSetup` for a given list of targets.
+///
+/// Every field type must implement [`BenchmarkingSetup`](crate::BenchmarkingSetup).
+///
+/// ```rust
+///
+/// struct Transfer;
+/// impl BenchmarkingSetup for Transfer { ... }
+///
+/// struct SetBalance;
+/// impl BenchmarkingSetup for SetBalance { ... }
+///
+/// selected_benchmark!(Transfer, SetBalance);
+/// ```
+#[macro_export]
+macro_rules! selected_benchmark {
+	($($bench:ident),*) => {
+		// The list of available benchmarks for this pallet.
+		enum SelectedBenchmark {
+			$( $bench, )*
+		}
+
+		// Allow us to select a benchmark from the list of available benchmarks.
+		impl<T: Trait> BenchmarkingSetup<T, crate::Call<T>, RawOrigin<T::AccountId>> for SelectedBenchmark {
+			fn components(&self) -> Vec<(BenchmarkParameter, u32, u32)> {
+				match self {
+					$( Self::$bench => <$bench as BenchmarkingSetup<T, crate::Call<T>, RawOrigin<T::AccountId>>>::components(&$bench), )*
```

Comment:
I'm a bit of a novice with this stuff myself, but I think this is now assuming that this macro is called in a context where `crate::Call<T>` is something. I think we have better workaround for these in our `decl_xxx` macros.

Reply:
Mmm, I think there are quite a few things wrong with this macro. I'll try next to put the whole thing inside a macro.